poly/src/polynomial/mod.rs

```diff
@@ -1,164 +1,93 @@
-        } else if divisor.is_zero() {
-            panic!("Dividing by zero polynomial")
-        } else if self.degree() < divisor.degree() {
-            Some((DensePolynomial::zero(), self.clone().into()))
-        } else {
-            // Now we know that self.degree() >= divisor.degree();
-            let mut quotient = vec![F::zero(); self.degree() - divisor.degree() + 1];
-            let mut remainder: DensePolynomial<F> = self.clone().into();
-            // Can unwrap here because we know self is not zero.
-            let divisor_leading_inv = divisor.leading_coefficient().unwrap().inverse().unwrap();
-            while !remainder.is_zero() && remainder.degree() >= divisor.degree() {
-                let cur_q_coeff = *remainder.coeffs.last().unwrap() * &divisor_leading_inv;
-                let cur_q_degree = remainder.degree() - divisor.degree();
-                quotient[cur_q_degree] = cur_q_coeff;
-
-                for (i, div_coeff) in divisor.iter_with_index() {
-                    remainder[cur_q_degree + i] -= &(cur_q_coeff * &div_coeff);
-                }
-                while let Some(true) = remainder.coeffs.last().map(|c| c.is_zero()) {
-                    remainder.coeffs.pop();
-                }
-            }
-            Some((DensePolynomial::from_coefficients_vec(quotient), remainder))
-        }
-    }
-}
-impl<'a, F: 'a + FftField> DenseOrSparsePolynomial<'a, F> {
-    /// Construct `Evaluations` by evaluating a polynomial over the domain
-    /// `domain`.
-    pub fn evaluate_over_domain<D: EvaluationDomain<F>>(
-        poly: impl Into<Self>,
-        domain: D,
-    ) -> Evaluations<F, D> {
-        let poly = poly.into();
-        poly.eval_over_domain_helper(domain)
-    }
-
-    fn eval_over_domain_helper<D: EvaluationDomain<F>>(self, domain: D) -> Evaluations<F, D> {
-        match self {
-            SPolynomial(Cow::Borrowed(s)) => {
-                let evals = domain.elements().map(|elem| s.evaluate(elem)).collect();
-                Evaluations::from_vec_and_domain(evals, domain)
-            }
-            SPolynomial(Cow::Owned(s)) => {
-                let evals = domain.elements().map(|elem| s.evaluate(elem)).collect();
-                Evaluations::from_vec_and_domain(evals, domain)
-            }
-            DPolynomial(Cow::Borrowed(d)) => {
-                Evaluations::from_vec_and_domain(domain.fft(&d.coeffs), domain)
-            }
-            DPolynomial(Cow::Owned(mut d)) => {
-                domain.fft_in_place(&mut d.coeffs);
-                Evaluations::from_vec_and_domain(d.coeffs, domain)
-            }
-        }
-    }
+    /// Given some point `z`, compute the quotients `w_i(X)` s.t
+    ///
+    /// `p(X) - p(z) = (X_1-z_1)*w_1(X) + (X_2-z_2)*w_2(X) + ... + (X_l-z_l)*w_l(X)`
+    ///
+    /// These quotients can always be found with no remainder.
+    fn divide_at_point(&self, point: &Self::Domain) -> Vec<Self>
```

Comment:
Can this be made into helper methods in `poly-commit`, instead of being a required method on the trait?